src/libraries/System.Net.Security/src/System/Net/Security/SslStream.Implementation.cs

```diff
@@ -305,6 +305,59 @@ private async Task ReplyOnReAuthenticationAsync<TIOAdapter>(TIOAdapter adapter,
             }
         }
 
+        // This will initiate renegotiation or PHA for Tls1.3
+        private async Task RenegotiateAsync(CancellationToken cancellationToken)
+        {
+            if (Interlocked.Exchange(ref _nestedAuth, 1) == 1)
+            {
+                throw new InvalidOperationException(SR.Format(SR.net_io_invalidnestedcall, "NegotiateClientCertificateAsync", "renegotiate"));
+            }
+
+            if (Interlocked.CompareExchange(ref _nestedRead, 1, 0) == 1)
+            {
+                throw new NotSupportedException(SR.Format(SR.net_io_invalidnestedcall, nameof(SslStream.ReadAsync), "read"));
+            }
+
+            if (Interlocked.CompareExchange(ref _nestedWrite, 1, 0) == 1)
+            {
+                _nestedRead = 0;
+                throw new NotSupportedException(SR.Format(SR.net_io_invalidnestedcall, nameof(WriteAsync), "write"));
+            }
+
+            _sslAuthenticationOptions!.RemoteCertRequired = true;
+            IReadWriteAdapter adapter = new AsyncReadWriteAdapter(InnerStream, cancellationToken);
+
+            try
+            {
+                SecurityStatusPal status = _context!.Renegotiate(out byte[]? nextmsg);
+                if (nextmsg?.Length > 0)
+                {
+                    await adapter.WriteAsync(nextmsg, 0, nextmsg.Length).ConfigureAwait(false);
+                    await adapter.FlushAsync().ConfigureAwait(false);
+                }
+
+                if (status.ErrorCode != SecurityStatusPalErrorCode.OK)
+                {
+                    if (status.ErrorCode == SecurityStatusPalErrorCode.NoRenegotiation)
+                    {
+                        // peer does not want to renegotiate. That should keep session usable.
+                        return;
+                    }
+
+                    throw SslStreamPal.GetException(status);
+                }
+
+                // Issue empty read to get renegotiation going.
+                await ReadAsyncInternal(adapter, Memory<byte>.Empty, renegotiation: true).ConfigureAwait(false);
```

Comment:
Why is this even necessary? It's not guaranteed to complete the renegotiation, is it?

Reply:
In fact, if I understand the logic here, isn't it going to potentially end up throwing away valid data?

Reply:
no, it is not guaranteed. I would need to go read the spec if client is even required to answer back. However _if_ the renegotiating starts, it would finish on Windows as it loops in `ForceAuthenticationAsync`.
So at this point it is best effort when we may end up with no remote certificate. 
We can also throw and about the SslStream if the client does not send back what we expect. We discuss this with @Tratcher while back as this is tricky to find right time when there are no other pending data.

Reply:
Renegotiating while the client is sending data inherently risks a deadlock if that data isn't being consumed. E.g. Checking cert auth on a POST request before reading the request body is a common problem people are going to run into. Making it fail with a clear error is one step better than having it deadlock. It _must not_ throw away valid data to give the appearance of success.

Reply:
I'm still trying to figure out what would be best way to handle this. Both Windows and Linux API sort of assume somebody will keep reading from the stream and that cranks the state machine. Since we blocked the actual SslStream.Read*(), this serves the purpose behind the scene.
One way would be to set time/size limits (or use cancellation) and fail the task if the renegotiation does not finish. That may or may not leave the Stream in usable form. (I think it would work or there would be clear error while doing IO later)

Reply:
I'm trying to understand how we expect this to work.

It seems like NegotiateClientCertificateAsync needs to *not* complete until the renegotiation is completed (or failed) and the cert is available. Otherwise you can't know when the cert is available. Correct?

At the same time, we need to be actively reading from the underlying stream, in order to receive the renegotiated certificate. I think this means that the *consumer* of the SslStream (i.e. Kestrel) needs to keep reading (or kill the stream completely). Is that correct?

The alternative would be for us to buffer a potentially really large amount of data in SslStream itself, until we receive the response to the renegotiation request. We'd need to set some limit here, or let the user specify the limit. We'd also need to keep processing (i.e. decrypting) the frames as we go, since we have to process frames in order. This seems messy.

Reply:
It is messy and historically problematic. @Tratcher was suggesting we would throw if we get data instead of renegotiation and it would be up to the caller to call it at correct time e.g. with knowledge of HTTP state. 
If we make caller to pump data, then we have two API calls that needs coordination and that seems problematic as well.

Reply:
NegotiateClientCertificateAsync should only be called when no data is expected (e.g. after the request has been received). NegotiateClientCertificateAsync should fail if any data is received before the client certificate. Buffering is a non-goal. Kestrel will not be reading from the stream during NegotiateClientCertificateAsync.

Reply:
Interesting... that certainly seems like the simplest approach. If that's a reasonable restriction (and it sounds like it is) then by all means, let's do that.

Reply:
It also seems like the safest approach, which I think was your point above. No chance of deadlock.

Reply:
ReadAsyncInternal will now throw if data are pending or received. With that the ReadAsyncInternal() will either fail of finish when renegotiation is completed.
We will need to verify with AAD that this meets their needs once available in Kestrel.